cmd/snap/cmd_set_test.go

```diff
@@ -127,6 +127,203 @@ func (s *snapSetSuite) TestSnapSetFailParsingWithStrictJSON(c *check.C) {
 	c.Assert(err, check.ErrorMatches, "failed to parse JSON:.*")
 }
 
+func (s *snapSetSuite) TestSnapSetJSONValidation(c *check.C) {
+
+	invalidJson := []string{
+		`key={"key0":"val0",:"val1","key2":{"key3":"val2","key4":"val3","key5":{"key6":"val4","key7":"val5"}},"key8":{"key9":{"key10":"val6"}}}`,
+		`key={"key0":"val0","key1":"val1","key2"{"key3":"val2","key4":"val3",` +
+			`"key5":{"key6":"val4","key7":"val5"}},"key8":{"key9":{"key10":"val6"}},` +
+			`"key11":[{"key12":"val7"},{"key13":"val8"},{"key14":["val9","val10"]}]}`,
+		`key={"key0": ["a", "b" "c"]}`,
+		`key=[{"key1": "a"} {"key2": "b"}]`,
+		`key=[{"key1": "a"}, {"key2": "b}]`,
+		`key={"key1": [123, "abc"}`,
+		`key={"key1": {"key2": "a", "key2",: ["b"]}}`,
+		`key={"key": {"key1": ,"value"}}`,
+	}
+
+	for _, json := range invalidJson {
+		_, err := snapset.Parser(snapset.Client()).ParseArgs([]string{"set", "snapname", "-t", json})
+		c.Assert(err, check.ErrorMatches, "failed to parse JSON:.*")
+	}
+
+	invalidJsonKeys := []string{
+		`key={"key0":"val0","key1":"val1","key2":{"key3":"val2","key4":"val3","key5":{"INVALID":"val4","key7":"val5"}},"key8":{"key9":{"key10":"val6"}}}`,
+		`key={"key0":"val0","key1":"val1","key2":{"key3":"val2","key4":"val3",` +
+			`"key5":{"key6":"val4","key7":"val5"}},"key8":{"key9":{"key10":"val6"}},` +
+			`"key11":[{"key12":"val7"},{"KEY13":"val8"},{"key14":["val9","val10"]}]}`,
+		`key={"KEY0": ["a", "b", "c"]}`,
+		`key=[{"key1": "a"}, {"KEY2": "b"}]`,
+		`key={"KEY1": [123, "abc"]}`,
+		`key={"key1": {"key2": "a", "KEY2": ["b"]}}`,
+		`key={"key1": [[123, 456], [789]]}`,
+		`key={"key1": "a", "key1": "b"}`,
```

Comment:
Also no invalid keys here, but the structure might be invalid (or gets the mysterious "access denied"), so it should probably be in `invalidJson` above instead.

BUT! Using this as a test case to debug the `"access denied"` error, if we change one of these keys to `KEY1` instead, then we get an `invalid option name:.*` error, as we want. So, we know that the `"access denied"` error happens AFTER key validation.

Reply:
I've tested locally, and in this case: `key={"key1": "a", "key1": "b"}`

The value store in `key` is:

```console
$ sudo /tmp/build/snap get -d envtester key                             
{
	"key": {
		"key1": "b"
	}
}
```
Which means: The last value overwrites the repeated keys. 

The access denied error didn't happen anymore, since the code change to other place, seems that it was related to the old test mocking code on `cmd/snap/`.